### DISCUSSION TITLE
[Proposal] Remove automatic UI thread marshaling from INotifyCollectionChanged handlers in Forms

### DISCUSSION THREAD

**Discussion by hartez** [xamarin/Xamarin.Forms#11823] [General]
**UPDATE**

If you have not read the background information provided in this post AND the discussion of PR #5866, stop. Hit the 'Cancel' button below the reply you're typing. Read this post, read the discussion thread of PR #5866, and then come back.

**END UPDATE**

This proposal is a breaking change for some users.

In all cases where Forms automatically marshals handling of the INotifyCollectionChanged interface's CollectionChanged event to the UI thread, we instead should throw an exception if the event is raised on a non-UI thread.

## Background

In all of the platforms supported by Xamarin.Forms, any changes to the UI have to be made on the main/UI thread. Trying to make a UI change from a non-UI thread (e.g., a thread from the thread pool) will result in some version of a cross-thread exception. 

In .NET, we often bind controls with multiple items to collections which implement the INotifyCollectionChanged interface. The interface contains a CollectionChanged event which the control can listen for. When the event is raised, the control can update itself to reflect the change in the underlying data. 

These two concepts collide when changes are made to one of these collections from a non-UI thread. When the data changes in a non-UI thread, the CollectionChanged event is raised and handled on that same non-UI thread; if those changes would result in a control change which modifies the UI (as most of them do), then those UI changes are happening off of the UI thread. A cross-thread exception ensues. 

There are two options to avoid this problem:

1. Make your collection changes on the UI thread. 
2. Handle the CollectionChanged event on the UI thread. 

Option 1 is the better option. Option 2 introduces some subtle problems, as we'll see later.

A long, long time ago (so far back that I can't link the commit because it's in the pre-open-source repo), a well-intentioned change was made to the ListProxy class (which handles a lot of bookkeeping for the ListView control). The most recent version of the change can be found [here](https://github.com/xamarin/Xamarin.Forms/blob/719fc7a604ff0cce8922d717c99bfb0fa17e35e0/Xamarin.Forms.Core/ListProxy.cs#L196). This code is basically automating option 2. The handler for the CollectionChanged event is running the update action using the Dispatcher. The Dispatcher checks to see if it's on the UI thread; if so, it immediately runs the update action. If not, it queues up the action on the UI thread. In the latter case, the action won't _actually_ be executed until the UI thread can get to it, which might be immediately or it might be _at some point in the future_. 

The intent of this change was to make the ListView easier to use; if you doing something off the main thread (say, processing a bunch of data) and needed to update the collection, you could safely do so - the actual update to the UI would be marshaled to the UI thread by ListProxy, and you wouldn't get a cross-thread exception. 

At the time, this was the only control in Forms which performed automatic marshaling when bound to an INotifyCollectionChanged source. Other controls notably _did not_ do this. The Pins collection for Xamarin.Forms.Maps, for example, required you to handle the marshaling yourself: see the discussion in [PR 5866](https://github.com/xamarin/Xamarin.Forms/pull/5866). 

After a whole lot of discussion, we decided to make the other INotifyCollectionChanged sources in Forms consistent with ListView. So we merged [PR 5866](https://github.com/xamarin/Xamarin.Forms/pull/5866) (for the Map Pins) and [PR 8568](https://github.com/xamarin/Xamarin.Forms/pull/8568) (for CollectionView). The intent was to make working with these collections/bindings off of the UI thread as simple and easy as possible. We thought we were improving the product.

We were wrong.

## Why the current solution does not work

The problem occurs under the following conditions:

1. The data changes.
2. An update to the control is queued up on the main thread. It will run later.
3. The control does something which depends on the current state of the data. 
4. The queued update on the main thread is processed.

It's during part 3 that things go pear-shaped. The control has not yet been notified that the underlying data has been changed; it has expectations about the shape of the data (e.g., number of items, order of items) which are no longer valid. When it tries to query the underlying data, the results are not in the expected shape and the control has to throw a consistency exception.

An example of this is [issue 10735](https://github.com/xamarin/Xamarin.Forms/issues/10735). The Android RecyclerView is attempting to animate some data updates, but the underlying data has already been changed; it's querying for data based on an index which is no longer valid, and an IndexOutOfBoundsException is thrown. [Issue 9753](https://github.com/xamarin/Xamarin.Forms/issues/9753) is a similar case. 

The problem is the gap between the update of the actual underlying data and the notification to the native control (the RecyclerView or ListView or UICollectionView) that the data has changed. Because the updates have to be queue on the UI thread, they are sometimes delayed, and during that delay any work the control does which queries the underlying data will get inconsistent results.

So simply pushing the handling of CollectionChanged onto the UI thread is a bust; it'll work a lot of the time, but when it fails it fails hard and in ways which are difficult to reproduce/diagnose. 

In the end, this means that the only way to safely update the data in a bound observable collection (INotifyCollectionChanged) is to do so on the UI thread (i.e., option 1 above). 

Thus, the proposal to remove all automatic marshaling for these collections. 

## What things look like if we don't accept this proposal

We currently have a workaround for this on Android in the form of PR 11235's [MarshalingObservableCollection](https://github.com/xamarin/Xamarin.Forms/pull/11235/files#diff-ac9ef8bd2cf50b8f95b122fbc71af02b), which automatically wraps any INCC used as a CollectionView ItemsSource on that platform. (To be clear - this problem is not limited to Android; all the platforms have it. Android is just the most common problem area right now, and we may apply MOC to the other platforms in the future). MOC basically watches for off-UI-thread changes in the wrapped collection and presents a snapshot of the pre-modification data to the CollectionView until the notifications can be processed. In short, it lies.

While this works to fix the two reported issues, we don't know if it covers all possibilities. And it adds additional overhead (in both memory allocation and CPU cycles) for _everyone_ using CollectionView, even if they aren't adding items off of the UI thread. And it adds additional maintenance overhead for Forms (and going forward, .NET MAUI).

So the trade-off is "more maintenance and less performance" for "some devs can add collection items from thread pool threads without having to call BeginInvokeOnMainThread". Keep in mind that we cannot even guarantee that the continued behavior won't run into further bugs.

## What things look like if we accept this proposal

From a maintenance perspective for Forms and .NET MAUI, things get simpler. We can drop some classes and their tests, and we have an entire class of bugs we don't have to worry about anymore. 

The trade off is that those devs who are relying on the current behavior (that they can add items to a collection off of the main thread without cross-thread exceptions) would have to change their code. The change would be from something like this:

```
// From a non-UI thread
myCollection.Add("item");
```

to this:

```
// From a non-UI thread
Device.BeginInvokeOnMainThread(() => myCollection.Add("item"));
```

In an attempt to surface this particular problem to the developer as soon as possible, at each location where we have been marshaling the event handler to the UI thread, we will instead throw an exception with very specific language regarding the change and how to address it.

**Comment by mattleibow** [comment]
How do the other data binding platforms work? I think that I usually have to marshal myself for say UWP?

**Reply by hartez** [reply]
Yes, you do.

**Comment by adammeaney** [comment]
I would ask what we believe the real replacement is to this, because I don't believe that anyone ever should write 

`Device.BeginInvokeOnMainThread(() => myCollection.Add("item"));`

In their VM. Forms code does not belong in your ViewModel.

Are we suggesting everyone wraps that in an IMainThread interface in their project and injects it into their ViewModel?

Are we suggesting we use SynchronizationContext and get the one from the UI thread and inject that?

**Reply by mackayn** [reply]
Use that approach already with IDeviceService in Prism so the viewmodel can be unit tested and that service mocked via  base view model. I'd prefer Forms is consistent and doesn't try and marshal to the UI thread for you. 

Agree the vm should not reference Xamarin.Forms if possible.

**Reply by hartez** [reply]
> I would ask what we believe the real replacement is to this, because I don't believe that anyone ever should write
> 
> `Device.BeginInvokeOnMainThread(() => myCollection.Add("item"));`
> 
> In their VM. Forms code does not belong in your ViewModel.

I'm not suggesting that you should put any of this code into your view model. Or at least, I didn't intend to suggest that. I'm simply saying that if you have an `INotifyCollectionChanged` in your view model, any changes to that collection should be made from the UI thread. 

To be honest, that goes for all view model changes. Your view model is an abstraction of your view. It is, by its very nature, tied to the UI and thus the UI thread. A change to the properties of your view model is a change to your UI. So when working with a UI which requires that all modifications happen on the UI thread (which is most of them), it follows that you should never be making changes to the view model from another thread. (Forms lets you get away with this for properties and we tried to let you get away with it for collections, but, well, see the root post.)

So no, don't add `BeginInvokeOnMainThread()` to your view model. And if you've got code that's running on a non-UI thread and making view model changes, I would suggest changing that (at least if the part of the view model that you're changing is or affects an `INotifyCollectionChanged`).

**Reply by mackayn** [reply]
I really don't see the problem with this change, it just makes things clearer.

**Comment by Tommigun1980** [comment]


**Reply by adammeaney** [reply]
What is the "correct thread"? From my ViewModel's perspective, why would I need to have a collection changed on the main thread. The only reason to do it is because the framework requires it, and if the framework is the one requiring it, the framework can fix it.

From my viewmodel, I might not even know what thread I am on most of the time. I would say most of my ViewModel operations don't really care what thread they were on, except they put network requests on the background thread.

After I get back from that network operation, why would I bother to go back to the main thread? From my ViewModel, there is no reason to do it. Only because Xamarin wants me to do it.

And if Xamarin requires all things that touch the View to be on the View thread, why don't they make that happen instead of me?

What if Xamarin requires INCC on the main thread, and WPF (this is an example, dont know how it works there), which my VMs could also target, requires INPC on the main thread too. So now I need to wrap every single property change in invoke on main thread, even though Xamarin doesn't need it. 

Littering my VM with threading stuff to get to the main thread just because Xamarin needs it doesnt really feel like it should be my responsibility, but theirs.

**Reply by hartez** [reply]
> Only because Xamarin wants me to do it.
> The only reason to do it is because the framework requires it, and if the framework is the one requiring it, the framework can fix it.

It's not an arbitrary requirement we created. It's a fundamental requirement of the native platforms. AFAIK, every UI platform we currently target requires this. 

> And if Xamarin requires all things that touch the View to be on the View thread, why don't they make that happen instead of me?

We _do_ make that happen for bindable properties (though I'd argue we shouldn't). Most frameworks don't. Please read the background information in the PR #5866 discussion. 

And we technically _can_ handle it for you for these collections in many situations, but doing so comes at a cost in performance for _everyone_, even if they aren't doing work off of the main thread. 

> What if Xamarin requires INCC on the main thread, and WPF (this is an example, dont know how it works there), which my VMs could also target, requires INPC on the main thread too. So now I need to wrap every single property change in invoke on main thread, even though Xamarin doesn't need it.

Well, first off - WPF and Forms _do_ have the same requirements here. Also, WPF _does not_ do the marshaling for you, so if you wanted to use your view models across the two platforms the same concerns would apply.

But even assuming they didn't, this is only an issue if you are making view model changes off of the UI thread. If you never do that, you never have to worry about this issue. 

> From my viewmodel, I might not even know what thread I am on most of the time. I would say most of my ViewModel operations don't really care what thread they were on, except they put network requests on the background thread.

Think about that statement for a bit. Your view model obviously _does_ care what thread it's on if you put your network requests on "the background thread". You wouldn't bother to do that if you didn't care about blocking the UI thread. 

Anyway, if the concern is about changes which may come from these background network calls, you just need to marshal the callback onto the UI thread. No need to wrap every VM property.

**Comment by programmation** [comment]
I'm not convinced that forcing developers to marshall data updates to the UI thread in ViewModels will solve the underlying issue. The issue as described appears to be that the control is referencing out of date data when there's an update queued to the main thread but not yet processed, so to me it seems more logical that the fix should be applied to that mechanism once, in the control, rather than every developer having to create their own version of it for all time from here on out.

**Reply by hartez** [reply]
> rather than every developer having to create their own version of it for all time from here on out.

Developers who are not making modifications to `INotifyCollectionChanged` collections from threads _other than the UI thread_ don't have to do _anything_.

**Reply by programmation** [reply]
However, bindable properties are in Controls, and `IDispatcher` is in each Control, no? So how do we reference the relevant `IDispatcher` from our ViewModels?

**Reply by KPixel** [reply]
The "referencing" happens when you bind the observable viewModel to the view.

For example, today, when you set a binding in XAML or with `view.SetBinding(...)`, Xamarin.Forms implicitly captures the dispatcher.
So, every time you do something like `viewModel.MyProperty = MyValue`, behind the scene, Xamarin receives the `OnPropertyChanged` event and does: `if (dispatcher.IsInvokeRequired) dispatcher.BeginInvokeOnMainThread(updateUI()) else updateUI();`

That's why you can change the viewModel from any thread today.

The Xamarin team was trying to do the same thing for observable collections but it leads to hard-to-discover bugs.

By the way, although it works mostly well for properties, it is still inefficient: The whole binding implementation uses more memory and is slower, especially if you update many properties at the "same" time. So, most apps that never use background threads still pay this price.

**Reply by programmation** [reply]
That's interesting (thanks!) but I was trying to ask how to capture an `IDispatcher` into my ViewModel so I can use it to make INCC updates on the UI thread. The problem as I see it is that the `IDispatcher` isn't accessible to the ViewModel unless I can inject it somehow - but where do I get it from to do that?

**Reply by KPixel** [reply]
The answer is the same: Capture the dispatcher when you bind the viewModel to the view.

```C#
myCollectionView.ItemsSource = new MarshalingObservableCollection(myCollection, Dispatcher);
```

This is just an example.
I think a "better" solution is to assume that the viewModel should only be modified on the UI thread, then any background task must receive and use the dispatcher...

See the discussion below for more details.

**Reply by KPixel** [reply]
Here is what a background task could look like: ([based on this comment](https://github.com/xamarin/Xamarin.Forms/pull/5866#issuecomment-483868320))

```C#
public async Task GetData()
{
    var items = await _dataService.RetrieveData().ConfigureAwait(false);

    // Some heavy data processing here, that shouldn't be on the main thread so that I don't have stuttering UIs

    _dispatcher.Dispatch(() => { // Now it is safe to change the viewModel
		foreach (var item in items)
		{
			_viewModel.MyObservableList.Add(item);
		}
	});
}
```

Edit: The background task receives the dispatcher at the same time that it receives the viewModel. They always go hand-in-hand.
The same thing would happened if you gave the view itself to the background task; any attempt to access that view would need to be wrapped around `view.Dispatcher.Dispatch(...)`.

**Reply by programmation** [reply]
I still can't bring myself to like this very much. If I have an `IDispatcher` in my ViewModel, that gives it a dependency on Xamarin.Forms, which doesn't look like a great move forward to me.

**Reply by KPixel** [reply]
Please, look again at all the examples I gave. I never put the dispatcher in the ViewModel.

**Reply by KPixel** [reply]
Also, this proposal is to remove the **automatic** marshaling. Not to forbid it.

If this proposal is approved, it will still be possible to do automatic marshaling, you will just have to opt-into this behavior (for example, by using something like the `MarshalingObservableCollection` class).

Currently, we as users of Xamarin.Forms don't have a choice. Even if we write a simple app that doesn't do any background work, we still have to pay the price in CPU & memory usage, as well as potential bugs.

**Reply by programmation** [reply]
You've still got to have a Dispatcher somewhere outside the View layer though. I know that we have to do this right now, but IMO the performance penalty of checking threads inside XF's Controls is well worth it if I never have to worry about thread exceptions. As it happens I have a `Threads` module (which doesn't use `IDispatcher`) that does the marshalling to the UI thread, but I only use it because things crash if I don't. I use `await` for the same reason - with that keyword we now no longer have to worry (as much) about what exact thread our code runs on. If I `await` something I know the .NET system will run it on the thread pool (could be the main thread; I don't have to care) and return data to me either on the same thread or (maybe) a different one. I get to choose, but the default behaviour is such that I can write code that runs on multiple threads without having to manage the threads myself. Same with XF - I'd love to have a suite of Controls that "just work" regardless of what thread the updates to their bound data occurs on. I know that at present this is hard, but it's a hard problem that every single XF developer has to solve individually. One of the reasons I use XF is because I'm hoping that it will solve hard problems like this for me, rather than forcing me (and all 5000000 other XF devs) to code everything myself. So if it's too hard to do this in the Controls themselves, how about creating an add-on to the Binding layer that auto-marshals data updates to the UI thread for those of us that want it?

**Reply by KPixel** [reply]
Using `await` will continue to work fine, even after this proposal is approved.

For example:

```C#
// MyButton.Clicked += (s, e) => GetData();
// MyButton.Command = new Command(() => GetData()); // Same with XAML

public async Task GetData()
{
    var items = await _dataService.RetrieveData();

    foreach (var item in items)
    {
        _viewModel.MyObservableList.Add(item);
    }
}
```

This code will always run on the UI thread, so it is safe.
And the code inside `RetrieveData()` can move to a different thread without any side-effects.

Do you have another example in mind that you think will stop working?

**Reply by programmation** [reply]
I'm not saying that `await` is a problem. I'm saying that I like the relative simplicity of being able to use `await` and have the .NET system abstract away the thread management for me. Similarly, I'd prefer it if XF could enable my code to not have to worry about explicitly marshalling data structure updates onto the UI thread just because the UI control that's bound to the data structure requires updates on the UI thread. I'd prefer the marshalling to occur somewhere other than in my Services layer or ViewModels. My current preference is to put it in the Binding layer since it could be made optional there with (I think) the fewest breaking changes for all other code.

**Reply by KPixel** [reply]
The simplicity you are talking about will remain, if your code is also simple.
You don't have to worry about marshalling as long as your code stays on the UI thread.

Can you give me an example where you think that this proposal will break your code?

**Reply by programmation** [reply]
I use MvvmCross, which initialises its ViewModels using an asynchronous Initialize() call. Since basically anything can happen in there, which will almost certainly involve updates to ViewModel properties that are bound to Views, and the code may be running on a non-UI thread, the INPC / INCC that results from those updates may need to be marshalled to the UI thread. At the moment I do this myself, but I would prefer for the XF binding system to handle it for me.

**Reply by KPixel** [reply]
I would assume that MvvmCross itself never does any UI work outside of the UI thread, and never accesses your Views/ViewModels outside of the UI thread. So, you should be able to continue using it like before. (Please, let me know if that's not the case.)

So, we go back to your code: As long as you never leave the UI thread, you are fine :)
Is there a scenario where you must leave the UI thread and don't want to have to explicitly get back on it?

**Reply by programmation** [reply]
MvvmCross initialises each ViewModel by calling `await Initialize()` on it, so the code in that method might be run on a background thread. All changes to ViewModel properties made via the Initialize() method will issue INPC or INCC calls to bound controls, so those calls may therefore need to be marshalled to the UI thread. MvvmCross happens to have tools for marshalling data updates to the UI thread, precisely because of this. I've also got my own set of those tools. I'm advocating for XF to have UI thread marshalling automation in its binding layer as an optional feature so we don't all have to build it for ourselves.

**Reply by KPixel** [reply]
`await Initialize()` is fine. You will stay on the UI thread.

You would only get into trouble if you added `.ConfigureAwait(false)` somewhere...

**Reply by programmation** [reply]
After Initialize(), yes I'll be back on whatever thread I started on. But Initialize is a virtual method that you override as needed for each ViewModel. Therefore the code that runs in it may be running on a non-UI thread, issuing INPC/INCC on that thread, unless you explicitly marshall those INPC/INCC updates back to the UI thread yourself. Which I do, but I'd prefer not to have to.

**Reply by KPixel** [reply]
Please, post an example of code running in `Initialize()` that would be problematic.

For example, this code doesn't require any marshaling: ([based on the doc](https://www.mvvmcross.com/documentation/fundamentals/viewmodel-lifecycle#initialize))

```C#
public override async Task Initialize()
{
    await base.Initialize();

    MyBoundLabel = "Loading...";

    var items = await _dataService.RetrieveData();

    foreach (var item in items)
    {
        MyBoundObservableList.Add(item);
    }
    MyBoundLabel = "Done";
}
```

**Reply by mackayn** [reply]
> Please, post an example of code running in `Initialize()` that would be problematic.
> 
> For example, this code doesn't require any marshaling: ([based on the doc](https://www.mvvmcross.com/documentation/fundamentals/viewmodel-lifecycle#initialize))
> 
> ```cs
> public override async Task Initialize()
> {
>     await base.Initialize();
> 
>     MyBoundLabel = "Loading...";
> 
>     var items = await _dataService.RetrieveData();
> 
>     foreach (var item in items)
>     {
>         MyBoundObservableList.Add(item);
>     }
>     MyBoundLabel = "Done";
> }
> ```

Yeah, but it slower, setting RetrieveData to run of the UI thread runs faster, you then push changes back to the UI thread. I know that code is safe but I wouldn't write it like that.

**Reply by KPixel** [reply]
In practice, the code inside `RetrieveData()` will typically be a network call (to a web API) followed by a bit of code to deserialize the response.

Something like:

```C#
public async Task<ItemList> RetrieveData()
{
    var json = await _httpClient.GetStringAsync("https://example.com/api/item/list").ConfigureAwait(false);
    var items = System.Text.Json.JsonSerializer.Deserialize<ItemList>(json);
    return items;
}
```

The network call is the part that takes the most time by far (let's say, at least 50 milliseconds) and all that time will be spent off the UI thread. Then, the `Deserialize()` call will typically happen off the UI thread as well, and in any case only [take some nanoseconds](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-5/).

When we return the items to `Initialize()`, we will come back on the UI thread, and quickly add them to the list.

I think this implementation, without automatic marshaling, is actually the fastest one! If we had left the UI thread explicitly, then each, each call to `MyBoundObservableList.Add(item);` with automatic marshaling would result in a lot of code to move each item back to the UI thread individually.

Please, read this blog post for more details:

https://devblogs.microsoft.com/dotnet/configureawait-faq/#when-should-i-use-configureawaitfalse

**Reply by mackayn** [reply]
> In practice, the code inside `RetrieveData()` will typically be a network call (to a web API) followed by a bit of code to deserialize the response.
> 
> Something like:
> 
> ```cs
> public async Task<ItemList> RetrieveData()
> {
>     var json = await _httpClient.GetStringAsync("https://example.com/api/item/list").ConfigureAwait(false);
>     var items = System.Text.Json.JsonSerializer.Deserialize<ItemList>(json);
>     return items;
> }
> ```
> 
> The network call is the part that takes the most time by far (let's say, at least 50 milliseconds) and all that time will be spent off the UI thread. Then, the `Deserialize()` call will typically happen off the UI thread as well, and in any case only [take some nanoseconds](https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-5/).
> 
> When we return the items to `Initialize()`, we will come back on the UI thread, and quickly add them to the list.
> 
> I think this implementation, without automatic marshaling, is actually the fastest one! If we had left the UI thread explicitly, then each, each call to `MyBoundObservableList.Add(item);` with automatic marshaling would result in a lot of code to move each item back to the UI thread individually.
> 
> Please, read this blog post for more details:
> 
> https://devblogs.microsoft.com/dotnet/configureawait-faq/#when-should-i-use-configureawaitfalse

That's not my experience in optimizing Xamarin apps and I've read many articles on this topic but I think we'll just go round in circles on this point arguing. I have no problem with the proposed change, just makes things clearer.

**Reply by KPixel** [reply]
If you have the time, do post some links to these articles. I do want to learn more, if I am missing something.

**Comment by KPixel** [comment]
I agree with the idea that this behavior should not be implicit.
However, I would like a solution that supports multi-window apps: What if a collection is bound to two views on two different windows?

I'm currently working on that scenario, and I have already been [hit by a few bugs in X.F](https://github.com/xamarin/Xamarin.Forms/issues/11705) related to the assumption that there is only ever a single window.

BTW, Device.BeginInvokeOnMainThread() should be avoided for that reason.

**Reply by hartez** [reply]
I used `Device.BeginInvokeOnMainThread()` because that's what most folks are familiar with, but yes - technically you should be using an instance of `IDispatcher` which you can retrieve from any `BindableObject`. And if you're working with a multi-window project (e.g., on UWP), you _must_ use `IDispatcher` to ensure that the update happens for the correct window.

**Reply by KPixel** [reply]
Can you please clarify your answer to: What if a collection is bound to two views on two different windows?
In that scenario, you would have two different dispatchers. So, there is no "correct" window.

**Reply by KPixel** [reply]
Trying to answer myself: With regular bindings, the binding itself can capture its dispatcher.
But with: myCollectionView.ItemsSource = mySharedCollection, that is not possible.
Instead, we should do: myCollectionView.ItemsSource = new MarshalingObservableCollection(mySharedCollection);

Correct? Is there any downside/potential issue with this solution, aside for the fact that any changes is not visible on the UI thread right away?

By the way, I think that `MarshalingObservableCollection` should be updated to capture the dispatcher directly [instead of using Device.BeginInvokeOnMainThread()](https://github.com/xamarin/Xamarin.Forms/blob/ad4a1a7838689a86a7183ae0b2a02177ba51861e/Xamarin.Forms.Core/Items/MarshalingObservableCollection.cs#L62). (Edit: Already reported here #11880)

I see that it is currently only being used in [one case for Android](https://github.com/xamarin/Xamarin.Forms/blob/ad4a1a7838689a86a7183ae0b2a02177ba51861e/Xamarin.Forms.Platform.Android/CollectionView/ItemsSourceFactory.cs#L27). So, that's probably why that issue was never mentioned.

**Reply by hartez** [reply]
> I see that it is currently only being used in one case for Android. So, that's probably why that issue was never mentioned.

Yes, that is precisely why we haven't moved it to IDispatcher yet. And if this proposal is accepted, we won't have to, because it will go away.

**Reply by KPixel** [reply]
So, what would be the recommended solution for that scenario? (A collection is bound to two views on two different windows)

**Reply by hartez** [reply]
> So, what would be the recommended solution for that scenario? (A collection is bound to two views on two different windows)

I would recommend not doing that. 

If I'm understanding your question correctly, you have a single instance of a view model, and you are binding controls from two different windows to that single instance. I'm going to assume that you aren't doing any two-way binding, or if you are it's not to controls in both windows (since that would have other obvious problems). 

I would recommend having a separate view model instance for each view, and updating each instance on the appropriate UI thread. That way you avoid any unintentional (and potentially hard to debug) conflicts between the two threads. Of course, this depends a lot on what's actually in your view model, and your model, and how you update them.

**Reply by KPixel** [reply]
I see the ViewModel as an extension to the View itself. So, with that understanding, I agree that each view/window should have its own ViewModel.

Here, the hypothetical collection that I want to share is a static list. For example, let's say it is a game that allows people to open multiple instances, each in its own window. Then, the players can each play on their own window (with touch screens which support multiple parallel inputs). Then, we have a leaderboard page to display the highest scores.
So, all the open windows must have access to this list, and any time a player wins, that list is updated, which affects all the windows.
How do we share this list?

**Reply by hartez** [reply]
In that case I'd recommend that each time your static list is updated, you broadcast that change to all the interested parties (windows, controls, etc.) with an event aggregator (or whatever your preferred pub/sub solution is). The listeners can update the various view models which back your actual controls.

This decouples the native controls entirely (e.g., ListView or RecyclerView or UICollectionView) from your primary static list. Then you don't have to worry at all about inconsistencies between the native controls' internal bookkeeping and the static list. 

If you have multiple list controls in multiple windows (and thus on multiple threads) all accessing the same static list, you're pretty much guaranteed to have a thread-safety issue. For example:

1. Window B's ListView accesses the static list's Count, and gets 8
2. Window A's thread removes an item from the list
3. Window B's ListView tries to animate in the 8th item (which no longer exists).

Give them all their own copy, and you avoid issues like this.

**Reply by KPixel** [reply]
If I understand correctly, that's basically what the `MarshalingObservableCollection` class does, right?

Hence, my suggestion:

```C#
myCollectionView.ItemsSource = new MarshalingObservableCollection(mySharedCollection, Dispatcher);
```

Edit: Of course, updated to pass a dispatcher.

**Comment by mattleibow** [comment]
What is the correct way to marshal to the main thread? For example, lets say I press a button which downloads some data. This operation has a callback/event that fires on the download background thread.

And, this VM that I am in does not reference the Xamarin.Forms (or any UI) framework.

What is the recommendation?

**Reply by Tommigun1980** [reply]
I misread the original proposal, and completely agree with @mattleibow.

It should not be the caller's responsibility to invoke changes to collections to the UI thread. If a framework draws something from a collection, there is already a separation of concern there - code manipulating the collection does not know what it's going to be used for.

An excellent example of this is VMs - none of my VMs have any kind of UI code in there and are purely business logic. I don't want to break that by suddenly adding Xamarin specific thread handling code there, or add yet another layer.

Please don't do this change as the caller shouldn't even know where the properties it exposes are used. I don't know what good my VMs would even do at that point.

I can't even fathom the amount of code smell this would introduce, not to mention, once again, making VMs and separation of concern utterly impossible.

**Reply by adammeaney** [reply]
@hartez  Guess we are just saying that the official advice is wrap IDispatcher, give it to your viewmodels, and anything that is an event (INPC, INCC, maybe your own events the view is listening to for some reason) should be queued up using the wrapped IDispatcher.BeginInvokeOnMainThread?

**Reply by hartez** [reply]
> It should not be the caller's responsibility to invoke changes to collections to the UI thread. 

It is on most platforms. 

> An excellent example of this is VMs - none of my VMs have any kind of UI code in there and are purely business logic. I don't want to break that by suddenly adding Xamarin specific thread handling code there, or add yet another layer.

Do you operate on your view model off of the UI thread? If you don't, then this is of no concern to you. If you have code which modifies your view model _off of the UI thread_, then you will need to marshal _those calls_ onto the UI thread. You do not need to add _any_ thread handling code to the view model itself.

> I can't even fathom the amount of code smell this would introduce, not to mention, once again, making VMs and separation of concern utterly impossible.

I must not be conveying this point well: what we are talking about here is reverting a change which was made _8 months ago_. So either your code prior to version 4.3 of Forms already had the code smell you are worried about, or this issue does not affect you.

> Guess we are just saying that the official advice is wrap IDispatcher, give it to your viewmodels, and anything that is an event (INPC, INCC, maybe your own events the view is listening to for some reason) should be queued up using the wrapped IDispatcher.BeginInvokeOnMainThread?

No. That is absolutely _not_ the official advice. 

If, for some reason, you have code somewhere in your application 

1. which modifies a collection
2. and that collection implements INotifyCollectionChanged
3. and that collection is the ItemsSource for a control (Map.Pins, CollectionView, etc.)
4. and that code does not run on the UI thread

then _that code_ should be marshaled to the UI thread, using whatever means you prefer (I would _suggest_ IDispatcher). 

How you accomplish that is entirely up to you.

**Reply by hartez** [reply]
> a callback/event that fires on the download background thread

Presumably your callback modifies the collection? Then the callback should marshal that modification to the UI thread. 

Think of it this way - what would you do in that callback if you weren't using some form of loose coupling like MVVM? If that callback just changed the text on a UILabel directly by setting the Text property? You'd have to marshal that call with something like `BeginInvokeOnMainThread`. Same deal here.

Forms [conveniently lets you ignore that requirement](https://github.com/xamarin/Xamarin.Forms/pull/5866#issuecomment-484311900) _much of the time_. This proposed change is about rolling back a few particular instances of that convenience because they introduce other problems and hurt performance.

**Reply by adammeaney** [reply]
And in that rolling back of this convenience, you have argued that you should roll back that convenience on property changes too. 

> We do make that happen for bindable properties (though I'd argue we shouldn't). Most frameworks don't. Please read the background information in the PR #5866 discussion.

But the basic thrust of what I am saying in this thread is that you recommended we use IDispatcher to handle any call that should be happening on the main thread. But any way of finding the main thread and running on it is ok for the same solution. Is that not what you meant by that? 

I have never known or cared which thread started a ViewModel command or property update, just that I don't block it when doing things I know will be long running. Whether I started on a background thread or the UI thread doesn't matter to my View Model, except in the fact that all my INCC (and potentially my INPC, if you decide to undo that convenience as well as it *should be my responsibility*) need to be on the main thread now.

**Reply by hartez** [reply]
> And in that rolling back of this convenience, you have argued that you should roll back that convenience on property changes too.

I personally don't think we should auto-marshal property bindings, no. But that's not part of this proposal. This proposal is _only_ about `INotfiyCollectionChanged`. Basic binding expressions are not on the table here. 

The `INotifyCollectionChanged` situation is being treated separately because 

1. The auto-marshaling is a relatively recent change (prior to December 2019 it _only_ existed for ListView)
2. It causes some demonstrable issues with the underlying controls 
3. The solutions for those problems add a performance hit for everyone, not just the folks who are doing things off the UI thread
4. The solutions for those problems add a maintenance burden to Forms, because they are hard to diagnose/debug when they go wrong

So we're proposing to roll that change back.

> But the basic thrust of what I am saying in this thread is that you recommended we use IDispatcher to handle any call that should be happening on the main thread. But any way of finding the main thread and running on it is ok for the same solution. Is that not what you meant by that?

I'm trying to parse this paragraph, but I'm not getting it. If you need a cross-platform way of marshaling stuff to the UI thread, then yes, I recommend IDispatcher because it will work correctly in multi-window scenarios. But feel free to use whatever you like to marshal the call to the UI thread.

> I have never known or cared which thread started a ViewModel command or property update, just that I don't block it when doing things I know will be long running. 

Do you write a lot of code where you don't know which thread you're on? 

> and potentially my INPC, if you decide to undo that convenience as well

Again, INPC is not on the table. INCC only.

So it sounds like your preference is strongly toward "I am fine with a small performance hit as long as I never have to think about threads."

**Comment by KPixel** [comment]
@hartez I just saw the update in your initial post. I think that it is too much reading for everyone :)

I would suggest having a short example right from the start to illustrate the problem that this proposal is trying to solve. Maybe, you can copy the `GetData()` code here.

**Reply by mattleibow** [reply]
Looking at the code in the linked issues in the update section really helped.

**Comment by GiampaoloGabba** [comment]
if the result of this rollback is more performance & stability, less bug, more maintanability, and align with UWP/WPF behaviours this is a no-brainer yes to me.... We should not even speaking of this, the PROS are simply too big :)

Maybe we can maintain the auto-marshaling for collections as opt-in (or opt-out) and update the documentation with a clear statement that this behaviour will have performance and stability downsides (basically: use it at your own risk). With this we dont break any existing code and we can move one from those nasty bugs.
